server/index.js

```diff
@@ -25,30 +25,137 @@ app.get('/api/products', (req, res, next) => {
+      return res.status(200).json(cartItems.rows);
+    })
+    .catch(err => next(err));
+});
+
+app.post('/api/cart', (req, res, next) => {
+  const productId = parseInt(req.body.productId);
+
+  if (!productId || isNaN(productId) || productId < 0) {
+    next(new ClientError(`Expected a positive integer. ${productId} is not a invalid id.`, 400));
+    return;
+  }
+
+  const sqlSelectQuery = `select "price"
+                          from "products"
+                          where "productId" = $1`;
+  const selectValues = [productId];
+  db.query(sqlSelectQuery, selectValues)
+    .then(selectResult => {
+      let sqlInsertQuery = `insert into "carts" ("cartId", "createdAt")
+                            values (default, default)
+                            returning "cartId"`;
+      let insertValues = [];
+      if (req.session.cartId && req.session.cartId !== null) {
+        sqlInsertQuery = `select "cartId"
+                          from "carts"
+                          where "cartId" = $1`;
+        insertValues = [req.session.cartId];
+      }
+      if (!selectResult.rowCount) {
+        next(new ClientError(`There is no price listed for the item with id: ${productId}`, 400));
+      }
+      return db.query(sqlInsertQuery, insertValues)
+        .then(insertResult => {
+          return { cartId: insertResult.rows[0].cartId, price: selectResult.rows[0].price };
+        })
+        .catch(err => next(err));
+    })
+    .then(cartData => {
+      req.session.cartId = cartData.cartId;
+      const sqlInsertQuery = `insert into "cartItems" ("cartId", "productId", "price")
+                              values ($1, $2, $3)
+                              returning "cartItemId"`;
+      const insertValues = [cartData.cartId, productId, cartData.price];
+      return db.query(sqlInsertQuery, insertValues)
+        .then(insertResult => {
+          return { cartItemId: insertResult.rows[0].cartItemId };
+        })
+        .catch(err => next(err));
+    })
+    .then(insertResult => {
+      const sqlInsertQuery = `select "c"."cartItemId",
+                                 "c"."price",
+                                 "p"."productId",
+                                 "p"."image",
+                                 "p"."name",
+                                 "p"."shortDescription"
+                          from "cartItems" as "c"
+                          join "products" as "p" using ("productId")
+                          where "c"."cartItemId" = $1`;
+      const insertValues = [insertResult.cartItemId];
+      db.query(sqlInsertQuery, insertValues)
+        .then(insertResult => {
+          return {
+            cartItemId: insertResult.rows[0].cartItemId,
+            price: insertResult.rows[0].price,
+            productId: insertResult.rows[0].productId,
+            image: insertResult.rows[0].image,
+            name: insertResult.rows[0].name,
+            shortDescription: insertResult.rows[0].shortDescription
```

Comment:
If you have everything you need here, you can respond instead of adding another `.then()` in the chain.